model/histogram/convert_test.go

```diff
@@ -0,0 +1,200 @@
+				buckets: []ExpectedBucket{
+					{le: "1", val: 20},
+					{le: "2", val: 40},
+					{le: "3", val: 60},
+					{le: "+Inf", val: 60},
+				},
+				count: 60,
+				sum:   100,
+			},
+		},
+		{
+			name: "Empty Histogram",
+			nhcb: &Histogram{
+				CustomValues:    []float64{},
+				PositiveBuckets: []int64{},
+				Count:           0,
+				Sum:             0.0,
+			},
+			labels: labels.FromStrings("__name__", "test_metric"),
+			expected: ExpectedClassicHistogram{
+				buckets: []ExpectedBucket{
+					{le: "+Inf", val: 0},
+				},
+				count: 0,
+				sum:   0,
+			},
+		},
+		{
+			name: "Missing __name__ label",
+			nhcb: &Histogram{
+				CustomValues:    []float64{1, 2, 3},
+				PositiveBuckets: []int64{10, 20, 30},
+				Count:           60,
+				Sum:             100.0,
+			},
+			labels:    labels.FromStrings("job", "test_job"),
+			expectErr: true,
+		},
+		{
+			name:      "Unsupported histogram type",
+			nhcb:      nil,
+			labels:    labels.FromStrings("__name__", "test_metric"),
+			expectErr: true,
+		},
+		{
+			name: "Histogram with zero bucket counts",
+			nhcb: &Histogram{
+				CustomValues:    []float64{1, 2, 3},
+				PositiveBuckets: []int64{0, 10, 0},
+				Count:           10,
+				Sum:             50.0,
+			},
+			labels: labels.FromStrings("__name__", "test_metric"),
+			expected: ExpectedClassicHistogram{
+				buckets: []ExpectedBucket{
+					{le: "1", val: 0},
+					{le: "2", val: 10},
+					{le: "3", val: 10},
+					{le: "+Inf", val: 10},
+				},
+				count: 10,
+				sum:   50,
+			},
+		},
+		{
+			name: "Mismatched bucket lengths",
+			nhcb: &Histogram{
+				CustomValues:    []float64{1, 2},
+				PositiveBuckets: []int64{10, 20, 30},
+				Count:           60,
+				Sum:             100.0,
+			},
+			labels:    labels.FromStrings("__name__", "test_metric"),
+			expectErr: true,
+		},
+		{
+			name: "single series Histogram",
+			nhcb: &Histogram{
+				CustomValues:    []float64{1},
+				PositiveBuckets: []int64{10},
+				Count:           10,
+				Sum:             20.0,
+			},
+			labels: labels.FromStrings("__name__", "test_metric"),
+			expected: ExpectedClassicHistogram{
+				buckets: []ExpectedBucket{
+					{le: "1", val: 10},
+					{le: "+Inf", val: 10},
+				},
+				count: 10,
+				sum:   20,
+			},
+		},
+	}
+
+	labelBuilder := labels.NewBuilder(labels.EmptyLabels())
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var got ExpectedClassicHistogram
+			err := ConvertNHCBToClassicHistogram(tt.nhcb, tt.labels, labelBuilder, func(lbls labels.Labels, val float64) error {
+				switch lbls.Get("__name__") {
+				case tt.labels.Get("__name__") + "_bucket":
+					got.buckets = append(got.buckets, ExpectedBucket{
+						le:  lbls.Get("le"),
+						val: val,
+					})
+				case tt.labels.Get("__name__") + "_count":
+					got.count = val
+				case tt.labels.Get("__name__") + "_sum":
+					got.sum = val
+				default:
+					return errors.New("unexpected metric name")
+				}
+				return nil
```

Comment:
Again, I would just pack this into some []series struct without parsing, otherwise we might skip some important bug